servant-auth/servant-auth-server/src/Servant/Auth/Server/Internal.hs

```diff
@@ -68,3 +82,137 @@ instance ( n ~ 'S ('S 'Z)
          -> (AuthResult v, SetCookieList n)
          -> ServerT (AddSetCookiesApi n api) Handler
       go fn (authResult, cookies) = addSetCookies cookies $ fn authResult
+
+
+{-
+NewAuth is a "quick" PoC to have a more modular way of providing
+authentications and the checking thereof.
+
+In the current implementation, all of the 'auths' are checked one
+by one, and the first that is present is tried and will either
+be returned when successful, or when not, throw an error given
+the appropriate 'ErrorFormatter' or return a 'Left err' if the
+'Lenient' modifier is set.
+Only if NONE of the auths are present will it either throw an
+'err401' or, if the 'Optional' modifier is set, return a 'Nothing'.
+This error might also be customizeable if we make it be required
+from the context.
+-}
+
+data NewAuth (mods :: [Type]) (auths :: [Type]) (a :: Type)
+    deriving (Typeable)
+
+type NewAuthResult a = Maybe (Either Text a)
+
+instance
+    ( HasServer api ctxs
+    , HasContextEntry (MkContextWithErrorFormatter ctxs) ErrorFormatters
+    , SBoolI (FoldRequired mods)
+    , SBoolI (FoldLenient mods)
+    , AllAuth auths a
+    , HasContextEntry ctxs (AuthHandler Request (NewAuthResult a))
+    ) => HasServer (NewAuth mods auths a :> api) ctxs where
+    type ServerT (NewAuth mods auths a :> api) m =
```

Comment:
Would it make sense define the instance for `NewAuth mods (auth ': auths) a :> api` to statically ensure that _at least_ one auth mode is defined ? We could define a custom type error for the case where the list of auths is empty.

Reply:
Hmm, probably a good idea, yeah.